src/CoreWCF.Primitives/src/CoreWCF/IdentityModel/Tokens/WSSecurityJan2004.cs

```diff
@@ -0,0 +1,580 @@
+                        return;
+                    }
+                }
+
+                throw DiagnosticUtility.ExceptionUtility.ThrowHelperError(new InvalidOperationException(SR.Format(SR.StandardsManagerCannotWriteObject, keyIdentifierClause.GetType())));
+            }
+        }
+
+        protected abstract class KeyIdentifierStrEntry : StrEntry
+        {
+            protected const string EncodingTypeValueBase64Binary = SecurityJan2004Strings.EncodingTypeValueBase64Binary;
+            protected const string EncodingTypeValueHexBinary = SecurityJan2004Strings.EncodingTypeValueHexBinary;
+            protected const string EncodingTypeValueText = SecurityJan2004Strings.EncodingTypeValueText;
+
+            protected abstract Type ClauseType { get; }
+            protected virtual string DefaultEncodingType { get { return EncodingTypeValueBase64Binary; } }
+            public abstract Type TokenType { get; }
+            protected abstract string ValueTypeUri { get; }
+            protected bool EmitBspRequiredAttributes { get; }
+
+            protected KeyIdentifierStrEntry(bool emitBspRequiredAttributes)
+            {
+                EmitBspRequiredAttributes = emitBspRequiredAttributes;
+            }
+
+            public override bool CanReadClause(XmlDictionaryReader reader, string tokenType)
+            {
+                if (reader.IsStartElement(CoreWCF.XD.SecurityJan2004Dictionary.KeyIdentifier, CoreWCF.XD.SecurityJan2004Dictionary.Namespace))
+                {
+                    string valueType = reader.GetAttribute(CoreWCF.XD.SecurityJan2004Dictionary.ValueType, null);
+                    return (ValueTypeUri == valueType);
+                }
+                return false;
+            }
+
+            protected abstract SecurityKeyIdentifierClause CreateClause(byte[] bytes, byte[] derivationNonce, int derivationLength);
+
+            public override Type GetTokenType(SecurityKeyIdentifierClause clause)
+            {
+                return TokenType;
+            }
+
+            public override SecurityKeyIdentifierClause ReadClause(XmlDictionaryReader reader, byte[] derivationNonce, int derivationLength, string tokenType)
+            {
+                string encodingType = reader.GetAttribute(CoreWCF.XD.SecurityJan2004Dictionary.EncodingType, null);
+                if (encodingType == null)
+                {
+                    encodingType = DefaultEncodingType;
+                }
+
+                reader.ReadStartElement();
+
+                byte[] bytes;
+                if (encodingType == EncodingTypeValueBase64Binary)
+                {
+                    bytes = reader.ReadContentAsBase64();
+                }
+                else if (encodingType == EncodingTypeValueHexBinary)
+                {
+                    throw new NotImplementedException();
+                   // bytes = HexBinary.Parse(reader.ReadContentAsString()).Value;
+                }
```

Comment:
For the WCF client I ported the one file SoapHexBinary. Take a look at it [here](https://github.com/dotnet/wcf/blob/c60577cf130b640fe407dc6a14750eaabfe203b3/src/System.Private.ServiceModel/src/Internals/System/Runtime/Remoting/SoapHexBinary.cs).